<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>CMAPLE: cmaple::Alignment Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">CMAPLE<span id="projectnumber">&#160;1.0.0</span>
   </div>
   <div id="projectbrief">C++ MAximum Parsimonious Likelihood Estimation</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>cmaple</b></li><li class="navelem"><a class="el" href="classcmaple_1_1_alignment.html">Alignment</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classcmaple_1_1_alignment-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">cmaple::Alignment Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="alignment_8h_source.html">alignment.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-types" name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:aee3f54a67adb568735bbc1db96944e3b"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> { <br />
&#160;&#160;<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba4e2c0251b5ac83d386bc3ba19cf528e4">IN_FASTA</a>
, <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba70a7e3edeb35284e39c8f9fe7214fde1">IN_PHYLIP</a>
, <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80">IN_MAPLE</a>
, <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a>
, <br />
&#160;&#160;<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba2fed9cbb7e537d6a2976264543c712cc">IN_UNKNOWN</a>
<br />
 }</td></tr>
<tr class="separator:aee3f54a67adb568735bbc1db96944e3b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a5a429c944ebac9aef2a2e20094a5dc2b"><td class="memItemLeft" align="right" valign="top"><a id="a5a429c944ebac9aef2a2e20094a5dc2b" name="a5a429c944ebac9aef2a2e20094a5dc2b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>Alignment</b> ()</td></tr>
<tr class="memdesc:a5a429c944ebac9aef2a2e20094a5dc2b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default constructor. <br /></td></tr>
<tr class="separator:a5a429c944ebac9aef2a2e20094a5dc2b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a88ba338239e3863ced3d5dc3e63a350f"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#a88ba338239e3863ced3d5dc3e63a350f">Alignment</a> (std::istream &amp;aln_stream, const std::string &amp;ref_seq=&quot;&quot;, const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> format=<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a>, const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a> seqtype=<a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a>)</td></tr>
<tr class="memdesc:a88ba338239e3863ced3d5dc3e63a350f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor from a stream of an alignment in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format.  <br /></td></tr>
<tr class="separator:a88ba338239e3863ced3d5dc3e63a350f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a128c4d12df8c70419d75edb9fd7b8d58"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#a128c4d12df8c70419d75edb9fd7b8d58">Alignment</a> (const std::string &amp;aln_filename, const std::string &amp;ref_seq=&quot;&quot;, const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> format=<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a>, const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a> seqtype=<a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a>)</td></tr>
<tr class="memdesc:a128c4d12df8c70419d75edb9fd7b8d58"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor from an alignment file in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format.  <br /></td></tr>
<tr class="separator:a128c4d12df8c70419d75edb9fd7b8d58"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a36d8b2bbaf72d21162d6e273bb3674e1"><td class="memItemLeft" align="right" valign="top"><a id="a36d8b2bbaf72d21162d6e273bb3674e1" name="a36d8b2bbaf72d21162d6e273bb3674e1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>~Alignment</b> ()</td></tr>
<tr class="memdesc:a36d8b2bbaf72d21162d6e273bb3674e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor. <br /></td></tr>
<tr class="separator:a36d8b2bbaf72d21162d6e273bb3674e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d0cc7a3b1321c2f22351e8801a8890a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#a1d0cc7a3b1321c2f22351e8801a8890a">read</a> (std::istream &amp;aln_stream, const std::string &amp;ref_seq=&quot;&quot;, const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> format=<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a>, const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a> seqtype=<a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a>)</td></tr>
<tr class="memdesc:a1d0cc7a3b1321c2f22351e8801a8890a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read an alignment from a stream in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format.  <br /></td></tr>
<tr class="separator:a1d0cc7a3b1321c2f22351e8801a8890a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a669f8cd477943459407b7ed32c934816"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#a669f8cd477943459407b7ed32c934816">read</a> (const std::string &amp;aln_filename, const std::string &amp;ref_seq=&quot;&quot;, const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> format=<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a>, const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a> seqtype=<a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a>)</td></tr>
<tr class="memdesc:a669f8cd477943459407b7ed32c934816"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read an alignment from a file in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format.  <br /></td></tr>
<tr class="separator:a669f8cd477943459407b7ed32c934816"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7cc0a73cc105173130dac55d78786f43"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#a7cc0a73cc105173130dac55d78786f43">write</a> (std::ostream &amp;aln_stream, const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> &amp;format=<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80">IN_MAPLE</a>)</td></tr>
<tr class="memdesc:a7cc0a73cc105173130dac55d78786f43"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write the alignment to a stream in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format.  <br /></td></tr>
<tr class="separator:a7cc0a73cc105173130dac55d78786f43"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a793e2c9b076c92298d812e35aa073952"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classcmaple_1_1_alignment.html#a793e2c9b076c92298d812e35aa073952">write</a> (const std::string &amp;aln_filename, const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> &amp;format=<a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80">IN_MAPLE</a>, const bool overwrite=false)</td></tr>
<tr class="memdesc:a793e2c9b076c92298d812e35aa073952"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write the alignment to a file in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format.  <br /></td></tr>
<tr class="separator:a793e2c9b076c92298d812e35aa073952"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Class presents the input alignment </p>
</div><h2 class="groupheader">Member Enumeration Documentation</h2>
<a id="aee3f54a67adb568735bbc1db96944e3b" name="aee3f54a67adb568735bbc1db96944e3b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aee3f54a67adb568735bbc1db96944e3b">&#9670;&#160;</a></span>InputType</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">cmaple::Alignment::InputType</a></td>
        </tr>
      </table>
</div><div class="memdoc">
<p><a class="el" href="classcmaple_1_1_alignment.html">Alignment</a> format </p>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="aee3f54a67adb568735bbc1db96944e3ba4e2c0251b5ac83d386bc3ba19cf528e4" name="aee3f54a67adb568735bbc1db96944e3ba4e2c0251b5ac83d386bc3ba19cf528e4"></a>IN_FASTA&#160;</td><td class="fielddoc"><p>FASTA format </p>
</td></tr>
<tr><td class="fieldname"><a id="aee3f54a67adb568735bbc1db96944e3ba70a7e3edeb35284e39c8f9fe7214fde1" name="aee3f54a67adb568735bbc1db96944e3ba70a7e3edeb35284e39c8f9fe7214fde1"></a>IN_PHYLIP&#160;</td><td class="fielddoc"><p>PHYLIP format </p>
</td></tr>
<tr><td class="fieldname"><a id="aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80" name="aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80"></a>IN_MAPLE&#160;</td><td class="fielddoc"><p><a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format </p>
</td></tr>
<tr><td class="fieldname"><a id="aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5" name="aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5"></a>IN_AUTO&#160;</td><td class="fielddoc"><p>Auto detect </p>
</td></tr>
<tr><td class="fieldname"><a id="aee3f54a67adb568735bbc1db96944e3ba2fed9cbb7e537d6a2976264543c712cc" name="aee3f54a67adb568735bbc1db96944e3ba2fed9cbb7e537d6a2976264543c712cc"></a>IN_UNKNOWN&#160;</td><td class="fielddoc"><p>Unknown format </p>
</td></tr>
</table>

</div>
</div>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a88ba338239e3863ced3d5dc3e63a350f" name="a88ba338239e3863ced3d5dc3e63a350f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a88ba338239e3863ced3d5dc3e63a350f">&#9670;&#160;</a></span>Alignment() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">cmaple::Alignment::Alignment </td>
          <td>(</td>
          <td class="paramtype">std::istream &amp;&#160;</td>
          <td class="paramname"><em>aln_stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>ref_seq</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a>&#160;</td>
          <td class="paramname"><em>format</em> = <code><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a></code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a>&#160;</td>
          <td class="paramname"><em>seqtype</em> = <code><a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a></code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor from a stream of an alignment in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_stream</td><td>A stream of an alignment file </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_seq</td><td>A reference sequence (optional). If not specified, it will be read from the alignment (in MAPLE format) or automatically generated from the alignment (in FASTA or PHYLIP format) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">format</td><td>Format of the alignment (optional): IN_MAPLE, IN_FASTA, IN_PHYLIP, or IN_AUTO (auto detection) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">seqtype</td><td>Data type of sequences (optional): SEQ_DNA (nucleotide data), SEQ_PROTEIN (amino acid data), or SEQ_AUTO (auto detection) </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if any of the following situations occur.<ul>
<li>the alignment is empty or in an incorrect format</li>
<li>the sequences contain invalid states </li>
</ul>
</td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a128c4d12df8c70419d75edb9fd7b8d58" name="a128c4d12df8c70419d75edb9fd7b8d58"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a128c4d12df8c70419d75edb9fd7b8d58">&#9670;&#160;</a></span>Alignment() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">cmaple::Alignment::Alignment </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>aln_filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>ref_seq</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a>&#160;</td>
          <td class="paramname"><em>format</em> = <code><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a></code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a>&#160;</td>
          <td class="paramname"><em>seqtype</em> = <code><a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a></code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor from an alignment file in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_filename</td><td>Name of an alignment file </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_seq</td><td>A reference sequence (optional). If not specified, it will be read from the alignment (in MAPLE format) or automatically generated from the alignment (in FASTA or PHYLIP format) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">format</td><td>Format of the alignment (optional): IN_MAPLE, IN_FASTA, IN_PHYLIP, or IN_AUTO (auto detection) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">seqtype</td><td>Data type of sequences (optional): SEQ_DNA (nucleotide data), SEQ_PROTEIN (amino acid data), or SEQ_AUTO (auto detection) </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if any of the following situations occur.<ul>
<li>the alignment is empty or in an incorrect format</li>
<li>the sequences contain invalid states</li>
</ul>
</td></tr>
    <tr><td class="paramname">std::ios_base::failure</td><td>if the alignment file is not found </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a1d0cc7a3b1321c2f22351e8801a8890a" name="a1d0cc7a3b1321c2f22351e8801a8890a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1d0cc7a3b1321c2f22351e8801a8890a">&#9670;&#160;</a></span>read() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void cmaple::Alignment::read </td>
          <td>(</td>
          <td class="paramtype">std::istream &amp;&#160;</td>
          <td class="paramname"><em>aln_stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>ref_seq</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a>&#160;</td>
          <td class="paramname"><em>format</em> = <code><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a></code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a>&#160;</td>
          <td class="paramname"><em>seqtype</em> = <code><a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a></code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read an alignment from a stream in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_stream</td><td>A stream of an alignment file </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_seq</td><td>A reference sequence (optional). If not specified, it will be read from the alignment (in MAPLE format) or automatically generated from the alignment (in FASTA or PHYLIP format) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">format</td><td>Format of the alignment (optional): IN_MAPLE, IN_FASTA, IN_PHYLIP, or IN_AUTO (auto detection) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">seqtype</td><td>Data type of sequences (optional): SEQ_DNA (nucleotide data), SEQ_PROTEIN (amino acid data), or SEQ_AUTO (auto detection) </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if any of the following situations occur.<ul>
<li>the alignment is empty or in an incorrect format</li>
<li>the sequences contain invalid states </li>
</ul>
</td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a669f8cd477943459407b7ed32c934816" name="a669f8cd477943459407b7ed32c934816"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a669f8cd477943459407b7ed32c934816">&#9670;&#160;</a></span>read() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void cmaple::Alignment::read </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>aln_filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>ref_seq</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a>&#160;</td>
          <td class="paramname"><em>format</em> = <code><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3ba5b481bc6c5d186f4e2912ab9636411f5">IN_AUTO</a></code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5d">cmaple::SeqRegion::SeqType</a>&#160;</td>
          <td class="paramname"><em>seqtype</em> = <code><a class="el" href="classcmaple_1_1_seq_region.html#abf1b0f9b15ce096626ce0a753540ea5daf242412da09240c41e37298e9fe157b5">cmaple::SeqRegion::SEQ_AUTO</a></code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read an alignment from a file in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_filename</td><td>Name of an alignment file </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_seq</td><td>A reference sequence (optional). If not specified, it will be read from the alignment (in MAPLE format) or automatically generated from the alignment (in FASTA or PHYLIP format) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">format</td><td>Format of the alignment (optional): IN_MAPLE, IN_FASTA, IN_PHYLIP, or IN_AUTO (auto detection) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">seqtype</td><td>Data type of sequences (optional): SEQ_DNA (nucleotide data), SEQ_PROTEIN (amino acid data), or SEQ_AUTO (auto detection) </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if any of the following situations occur.<ul>
<li>the alignment is empty or in an incorrect format</li>
<li>the sequences contain invalid states</li>
</ul>
</td></tr>
    <tr><td class="paramname">std::ios_base::failure</td><td>if the alignment file is not found </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a7cc0a73cc105173130dac55d78786f43" name="a7cc0a73cc105173130dac55d78786f43"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7cc0a73cc105173130dac55d78786f43">&#9670;&#160;</a></span>write() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void cmaple::Alignment::write </td>
          <td>(</td>
          <td class="paramtype">std::ostream &amp;&#160;</td>
          <td class="paramname"><em>aln_stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> &amp;&#160;</td>
          <td class="paramname"><em>format</em> = <code><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80">IN_MAPLE</a></code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Write the alignment to a stream in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_stream</td><td>A stream of the output alignment file </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">format</td><td>Format of the output alignment (optional): IN_MAPLE, IN_FASTA, or IN_PHYLIP </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if the format is unknown </td></tr>
    <tr><td class="paramname">std::logic_error</td><td>if the alignment is empty (i.e., nothing to write) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a793e2c9b076c92298d812e35aa073952" name="a793e2c9b076c92298d812e35aa073952"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a793e2c9b076c92298d812e35aa073952">&#9670;&#160;</a></span>write() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void cmaple::Alignment::write </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>aln_filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3b">InputType</a> &amp;&#160;</td>
          <td class="paramname"><em>format</em> = <code><a class="el" href="classcmaple_1_1_alignment.html#aee3f54a67adb568735bbc1db96944e3baa1d0d04b8d6e30dc1340d981bf223e80">IN_MAPLE</a></code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>overwrite</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Write the alignment to a file in FASTA, PHYLIP, or <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_filename</td><td>Name of the output alignment file </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">format</td><td>Format of the output alignment (optional): IN_MAPLE, IN_FASTA, or IN_PHYLIP </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">overwrite</td><td>TRUE to overwrite the existing output file (optional) </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if any of the following situations occur.<ul>
<li>aln_filename is empty</li>
<li>format is unknown</li>
</ul>
</td></tr>
    <tr><td class="paramname">std::logic_error</td><td>if the alignment is empty (i.e., nothing to write) </td></tr>
    <tr><td class="paramname">std::ios_base::failure</td><td>if aln_filename already exists and overwrite = false </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>alignment/<a class="el" href="alignment_8h_source.html">alignment.h</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8
</small></address>
</body>
</html>
